compiler/rustc_resolve/src/late.rs

```diff
@@ -1703,10 +1703,28 @@ impl<'a: 'ast, 'b, 'ast, 'tcx> LateResolutionVisitor<'a, 'b, 'ast, 'tcx> {
                                 break;
                             }
                         }
-                        self.r.dcx().emit_err(errors::ElidedAnonymousLivetimeReportError {
-                            span: lifetime.ident.span,
-                            suggestion,
-                        });
+
+                        // Is it caused by user trying to implement a lending iterator?
+                        if !self.in_func_body
+                            && let Some((module, _)) = &self.current_trait_ref
+                            && let Some(ty) = &self.diag_metadata.current_self_type
+                            && let crate::ModuleKind::Def(DefKind::Trait, trait_id, _) = module.kind
+                            && def_id_matches_path(
+                                self.r.tcx,
+                                trait_id,
+                                &["core", "iter", "traits", "iterator", "Iterator"],
```

Comment:
Can this be extended to all non-generic associated types?

Reply:
Which part specifically?

Reply:
I think Jack is saying that this PR has been tailored to just `Iterator` (and `Iterator::Item`), and is then posing the thought exercise of whether the logic here (and the associated diagnostic output) could be generalized, so that it would cover all associated types lacking a parameter (i.e. "non-generic").

I.e., instead of specializing this check to `trait Iterator { type Item; ... }`, can it be extended to `trait <AnyTraitNameHere> { type AnyAssociatedTypeHere; ... // but not GATs }` ?

Reply:
Possibly, but I think this would be a separate diagnostic. I've seen people trying to implement lending iterators, I even tried to implement it myself years ago before I understood how it works :) So note about iterators is the main part and I want to keep it.

I can try to make the check more generic and add an error message about using non-GAT associated type.

Reply:
Rebased, added a better error message for general case of non-GAT associated type, left the error for iterators as is.

Reply:
And I'm open to any suggestions on wording/formatting/etc.